File: integrations/langchain-python/xache_langchain/reputation.py
"""
Xache Reputation for LangChain
Portable, verifiable agent reputation with ERC-8004 support
"""

from typing import Optional, Dict, Any
from langchain.tools import BaseTool
from pydantic import BaseModel, Field

from xache import XacheClient


class ReputationResult(BaseModel):
    """Reputation query result"""
    score: float = Field(description="Reputation score (0-1)")
    level: str = Field(description="Reputation level")
    total_contributions: int = Field(default=0)
    total_payments: int = Field(default=0)
    erc8004_enabled: bool = Field(default=False)
    erc8004_agent_id: Optional[str] = Field(default=None)


class XacheReputationTool(BaseTool):
    """
    LangChain tool for checking Xache reputation.

    Use this tool to check your agent's reputation score and status.
    High reputation unlocks benefits like lower prices and higher trust.

    Example:
        ```python
        from xache_langchain import XacheReputationTool

        rep_tool = XacheReputationTool(
            wallet_address="0x...",
            private_key="0x..."
        )

        # Use in agent
        tools = [rep_tool, ...]
        ```
    """

    name: str = "xache_check_reputation"
    description: str = (
        "Check your current reputation score and status. "
        "Returns your score (0-1), level, and ERC-8004 on-chain status. "
        "Higher reputation means lower costs and more trust from other agents."
    )

    # Xache configuration
    api_url: str = "https://api.xache.xyz"
    wallet_address: str
    private_key: str
    chain: str = "base"

    _client: Optional[XacheClient] = None

    class Config:
        arbitrary_types_allowed = True
        underscore_attrs_are_private = True

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        chain_prefix = "sol" if self.chain == "solana" else "evm"
        did = f"did:agent:{chain_prefix}:{self.wallet_address.lower()}"
        self._client = XacheClient(
            api_url=self.api_url,
            did=did,
            private_key=self.private_key,
        )

    def _run(self) -> str:
        """Check reputation"""
        import asyncio

        async def _check():
            async with self._client as client:
                result = await client.reputation.get_score()
                return result

        try:
            loop = asyncio.get_event_loop()
            if loop.is_running():
                import concurrent.futures
                with concurrent.futures.ThreadPoolExecutor() as pool:
                    result = pool.submit(asyncio.run, _check()).result()
            else:
                result = loop.run_until_complete(_check())
        except RuntimeError:
            result = asyncio.run(_check())

        score = result.get("score", 0)
        level = self._get_level(score)

        output = f"Reputation Score: {score:.2f}/1.00 ({level})\n"

        if result.get("erc8004AgentId"):
            output += f"ERC-8004 Status: Enabled (Agent ID: {result['erc8004AgentId']})\n"
            output += "Your reputation is verifiable on-chain!"
        else:
            output += "ERC-8004 Status: Not enabled\n"
            output += "Enable ERC-8004 to make your reputation portable and verifiable."

        return output

    async def _arun(self) -> str:
        """Async check reputation"""
        async with self._client as client:
            result = await client.reputation.get_score()

        score = result.get("score", 0)
        level = self._get_level(score)

        output = f"Reputation Score: {score:.2f}/1.00 ({level})\n"

        if result.get("erc8004AgentId"):
            output += f"ERC-8004 Status: Enabled (Agent ID: {result['erc8004AgentId']})\n"
        else:
            output += "ERC-8004 Status: Not enabled\n"

        return output

    def _get_level(self, score: float) -> str:
        """Get reputation level from score"""
        if score >= 0.9:
            return "Elite"
        elif score >= 0.7:
            return "Trusted"
        elif score >= 0.5:
            return "Established"
        elif score >= 0.3:
            return "Developing"
        else:
            return "New"


class XacheReputationChecker:
    """
    Utility class for checking reputation of any agent.

    Useful for verifying other agents before interacting.

    Example:
        ```python
        from xache_langchain import XacheReputationChecker

        checker = XacheReputationChecker(
            wallet_address="0x...",
            private_key="0x..."
        )

        # Check another agent's reputation
        other_rep = checker.check("did:agent:evm:0xOtherAgent...")
        if other_rep.score >= 0.5:
            print("Agent is trustworthy")
        ```
    """

    def __init__(
        self,
        wallet_address: str,
        private_key: str,
        api_url: str = "https://api.xache.xyz",
        chain: str = "base",
    ):
        self.api_url = api_url
        self.chain = chain

        chain_prefix = "sol" if chain == "solana" else "evm"
        self.did = f"did:agent:{chain_prefix}:{wallet_address.lower()}"

        self._client = XacheClient(
            api_url=api_url,
            did=self.did,
            private_key=private_key,
        )

    def check(self, agent_did: str) -> ReputationResult:
        """Check an agent's reputation"""
        import asyncio

        async def _check():
            async with self._client as client:
                result = await client.reputation.get_score(agent_did=agent_did)
                return result

        try:
            loop = asyncio.get_event_loop()
            if loop.is_running():
                import concurrent.futures
                with concurrent.futures.ThreadPoolExecutor() as pool:
                    result = pool.submit(asyncio.run, _check()).result()
            else:
                result = loop.run_until_complete(_check())
        except RuntimeError:
            result = asyncio.run(_check())

        score = result.get("score", 0)

        return ReputationResult(
            score=score,
            level=self._get_level(score),
            total_contributions=result.get("totalContributions", 0),
            total_payments=result.get("totalPayments", 0),
            erc8004_enabled=bool(result.get("erc8004AgentId")),
            erc8004_agent_id=result.get("erc8004AgentId"),
        )

    async def acheck(self, agent_did: str) -> ReputationResult:
        """Async check an agent's reputation"""
        async with self._client as client:
            result = await client.reputation.get_score(agent_did=agent_did)

        score = result.get("score", 0)

        return ReputationResult(
            score=score,
            level=self._get_level(score),
            total_contributions=result.get("totalContributions", 0),
            total_payments=result.get("totalPayments", 0),
            erc8004_enabled=bool(result.get("erc8004AgentId")),
            erc8004_agent_id=result.get("erc8004AgentId"),
        )

    def _get_level(self, score: float) -> str:
        """Get reputation level from score"""
        if score >= 0.9:
            return "Elite"
        elif score >= 0.7:
            return "Trusted"
        elif score >= 0.5:
            return "Established"
        elif score >= 0.3:
            return "Developing"
        else:
            return "New"

File: integrations/langchain-ts/src/reputation.ts
/**
 * Xache Reputation for LangChain.js
 * Portable, verifiable agent reputation with ERC-8004 support
 */

import { DynamicStructuredTool } from '@langchain/core/tools';
import { z } from 'zod';
import { XacheClient, DID } from '@xache/sdk';

export interface ReputationToolConfig {
  /** Wallet address for authentication */
  walletAddress: string;
  /** Private key for signing */
  privateKey: string;
  /** API URL (defaults to https://api.xache.xyz) */
  apiUrl?: string;
  /** Chain: 'base' or 'solana' */
  chain?: 'base' | 'solana';
}

export interface ReputationResult {
  /** Reputation score (0-1) */
  score: number;
  /** Reputation level */
  level: string;
  /** Total contributions made */
  totalContributions: number;
  /** Total payments made */
  totalPayments: number;
  /** Whether ERC-8004 is enabled */
  erc8004Enabled: boolean;
  /** ERC-8004 agent ID if enabled */
  erc8004AgentId?: string;
}

/**
 * Get reputation level from score
 */
function getLevel(score: number): string {
  if (score >= 0.9) return 'Elite';
  if (score >= 0.7) return 'Trusted';
  if (score >= 0.5) return 'Established';
  if (score >= 0.3) return 'Developing';
  return 'New';
}

/**
 * Create an Xache client for reputation tools
 */
function createClient(config: ReputationToolConfig): XacheClient {
  const chainPrefix = config.chain === 'solana' ? 'sol' : 'evm';
  const did = `did:agent:${chainPrefix}:${config.walletAddress.toLowerCase()}` as DID;

  return new XacheClient({
    apiUrl: config.apiUrl || 'https://api.xache.xyz',
    did,
    privateKey: config.privateKey,
  });
}

/**
 * Create a tool for checking your own reputation.
 *
 * @example
 * ```typescript
 * import { createReputationTool } from '@xache/langchain';
 *
 * const repTool = createReputationTool({
 *   walletAddress: '0x...',
 *   privateKey: '0x...',
 * });
 *
 * const tools = [repTool, ...];
 * ```
 */
export function createReputationTool(
  config: ReputationToolConfig
): DynamicStructuredTool {
  const client = createClient(config);

  return new DynamicStructuredTool({
    name: 'xache_check_reputation',
    description:
      'Check your current reputation score and status. ' +
      'Returns your score (0-1), level, and ERC-8004 on-chain status. ' +
      'Higher reputation means lower costs and more trust from other agents.',
    schema: z.object({}),
    func: async () => {
      const result = await client.reputation.getReputation();

      // Overall score is 0-100, normalize to 0-1
      const score = (result.overall || 0) / 100;
      const level = getLevel(score);

      let output = `Reputation Score: ${score.toFixed(2)}/1.00 (${level})\n`;
      output += `Memory Quality: ${result.memoryQuality || 0}/100\n`;
      output += `Contribution Success: ${result.contribSuccess || 0}/100\n`;
      output += `Economic Value: ${result.economicValue || 0}/100\n`;

      // Check ERC-8004 status separately
      try {
        const erc8004Status = await client.reputation.getERC8004Status();
        if (erc8004Status.enabled) {
          output += `ERC-8004 Status: Enabled\n`;
          output += 'Your reputation is verifiable on-chain!';
        } else {
          output += 'ERC-8004 Status: Not enabled\n';
          output += 'Enable ERC-8004 to make your reputation portable and verifiable.';
        }
      } catch {
        output += 'ERC-8004 Status: Unknown';
      }

      return output;
    },
  });
}

/**
 * Class-based reputation tool (alternative API)
 */
export class XacheReputationTool {
  private tool: DynamicStructuredTool;

  constructor(config: ReputationToolConfig) {
    this.tool = createReputationTool(config);
  }

  asTool(): DynamicStructuredTool {
    return this.tool;
  }
}

/**
 * Utility class for checking reputation of any agent.
 *
 * @example
 * ```typescript
 * import { XacheReputationChecker } from '@xache/langchain';
 *
 * const checker = new XacheReputationChecker({
 *   walletAddress: '0x...',
 *   privateKey: '0x...',
 * });
 *
 * const rep = await checker.check('did:agent:evm:0xOtherAgent...');
 * if (rep.score >= 0.5) {
 *   console.log('Agent is trustworthy');
 * }
 * ```
 */
export class XacheReputationChecker {
  private client: XacheClient;

  constructor(config: ReputationToolConfig) {
    this.client = createClient(config);
  }

  /**
   * Check an agent's reputation
   */
  async check(agentDid: string): Promise<ReputationResult> {
    const result = await this.client.reputation.getReputation(agentDid as DID);

    // Overall score is 0-100, normalize to 0-1
    const score = (result.overall || 0) / 100;

    // Try to get ERC-8004 status
    let erc8004Enabled = false;
    let erc8004AgentId: string | undefined;
    try {
      const erc8004Status = await this.client.reputation.getERC8004Status();
      erc8004Enabled = erc8004Status.enabled;
    } catch {
      // ERC-8004 status not available
    }

    return {
      score,
      level: getLevel(score),
      totalContributions: 0, // Not available in current API
      totalPayments: 0, // Not available in current API
      erc8004Enabled,
      erc8004AgentId,
    };
  }

  /**
   * Check if an agent meets minimum reputation threshold
   */
  async meetsThreshold(agentDid: string, minScore: number): Promise<boolean> {
    const result = await this.check(agentDid);
    return result.score >= minScore;
  }
}
